docs/tradeoffs.md

```diff
@@ -0,0 +1,136 @@
+  while giving the best build times and a reasonable to understand code with
+  the `derive` API
+- All other cases the answer is "it depends" as the two designs offer different
+  trade offs.   If you want to just choose one with little research that will
+  cover the most use cases, that will most likely be `clap`.
+
+## Static vs Dynamic Typing
+
+`bpaf`'s combinator API uses generics and macros to build a statically-typed
+parser, pushing development errors to compile time.  This minimizes binary
+size and runtime, only paying for what you use.
+
+The combinator approach tends towards yoda-speak though with [careful
+structuring by breaking down arguments into
+functions](https://github.com/pacak/bpaf/blob/aa6992931bbfbdca6390c87f4a76898f8db0ae47/examples/top_to_bottom.rs),
+a more straightforward ordering can be accomplished.
+
+`clap`'s builder API stores parsed values in a Map-like container where the
+keys are the argument IDs and the values are effectively `Box<dyn Any>`.
+- This allows dynamically created CLIs, like with
+  [clap_serde](https://docs.rs/clap_serde) or [conditionally-present
+  flags](https://github.com/sharkdp/bat/blob/6680f65e4b25b0f18c455f7a4639a96e97519dc5/src/bin/bat/clap_app.rs#L556)
+- Callers can dynamically iterate over the arguments, like for layered configs.
+- `clap` can have ready-to-use, arbitrarily deifned defaults and validation
+  conditioned on argument values (e.g. `default_value_if`).  See the section on
+  Validation for more nuance on the "arbitrarily" part
+- `clap` tries to help push as many errors to early test time with minimal
+  effort with the `cmd.debug_assert()` function.  Value lookups still requires
+  full coverage of that code to catch errors.
+
+Which design approach is faster to build will be dependent on the exact
+implementation and the compiler.
+
+## Context-sensitive parsing
+
+Parsing arguments is context sensitive.  Consider:
+```console
+$ prog --one two --three --four -abcdef
+```
+- When is an argument a subcommand, a positional value, a flag, or a flag's value?
+- When is the trailing parts of a short flag an attached value or more shorts flags?
+
+For example, some possible interpretations of the above could be:
+```console
+$ prog --one=two --three=--four -a=bcdef
+$ prog two -a -b -c -d -e -f --one --three --four
+```
+
+`clap` parses left-to-right using [`clap_lex`](https://docs.rs/clap_lex), with
+the output from the previous token hinting how to parse the next one to
+prevent ambiguity.  This leads to a fairly complicated parser to handle all of
+the different cases upfront, including some usability features to help catch
+developer mistakes.
+
+`bpaf` instead does context-free tokenization, storing all flags and values in
+a list.
+- Tokens that look like flags can only be considered flags, unless escaped with
+  `--`, not allowing cases like `fd --exec rm -rf ;`
+- Short flags can only have attached values in the `-a=b` case (not even `-ab=c`)
```

Comment:
If `-b` captures a value, all of the following are the same in clap:
- `-abc`
- `-ab=c`
- `-a -bc`
- `-a -b=c`
- `-a -b c`

Reply:
Hmm... What if `-c` can also capture a value and user passes `-abcd`? Anyway, I don't think I want to support it unless there's a real need for that.

Reply:
iirc `-c` wins and consumes `d` rather than parsing it.

iirc this is in contrast to our `allow_hyphen_values` which will give precedence to existing short/long flags but at least in that case there is a more clear syntax saying there is a potential flag (`--`) than with an attached value.